### PR TITLE
Checkout: validate empty cart state before placing an order.

### LINKED ISSUE #47967: [Blocks] Order can be placed even if all the cart items are deleted before
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

If you delete the cart items and then place the order, the order will still be successfully made.

https://github.com/woocommerce/woocommerce/assets/9415030/35fe1690-d2a0-47b0-875d-c1b3aa6bdf23

### Expected behavior

The order should not be placed. An error message should appear: "Your cart is empty".

### Actual behavior

The order is placed, and empty.

### Steps to reproduce

1. Go to a product page
2. Add it to cart
3. Open Cart page in a new tab
4. Open checkout page in a new tab
5. Fill in all the checkout fields but do no tplace the order
6. Go to the Cart tab and remove the cart item
7. Go to the Checkout page and place the order

<details>
<summary>WordPress Environment</summary>

```
### WordPress Environment ###

WordPress address (URL): http://test.booking-activities.yc
Site address (URL): http://test.booking-activities.yc
WC Version: 8.9.1
REST API Version: ✔ 8.9.1
Action Scheduler Version: ✔ 3.7.4
Log Directory Writable: ✔
WP Version: 6.5.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.54 (Win64) OpenSSL/1.1.1o PHP/8.1.7 mod_fcgid/2.3.10-dev
PHP Version: 8.1.7
PHP Post Max Size: 8 MB
PHP Time Limit: 60
PHP Max Input Vars: 2500
cURL Version: 7.77.0
OpenSSL/1.1.1o

SUHOSIN Installed: –
MySQL Version: 8.0.29
Max Upload Size: 8 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.9.1
WC Database Prefix: wp_
Total Database Size: 2.04MB
Database Data Size: 1.77MB
Database Index Size: 0.27MB
wp_woocommerce_sessions: Data: 0.01MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_api_keys: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_attribute_taxonomies: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_downloadable_product_permissions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_order_items: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_order_itemmeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_tax_rates: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_tax_rate_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zones: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_methods: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokens: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokenmeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_actionscheduler_actions: Data: 0.01MB + Index: 0.02MB + Engine MyISAM
wp_actionscheduler_claims: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_actionscheduler_groups: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_actionscheduler_logs: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_commentmeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_comments: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_links: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_options: Data: 1.67MB + Index: 0.04MB + Engine MyISAM
wp_postmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_posts: Data: 0.02MB + Index: 0.01MB + Engine MyISAM
wp_term_relationships: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_term_taxonomy: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_termmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_terms: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_usermeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_users: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.01MB + Engine MyISAM
wp_wc_admin_notes: Data: 0.03MB + Index: 0.00MB + Engine MyISAM
wp_wc_category_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_customer_lookup: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_download_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_order_addresses: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_order_coupon_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_order_operational_data: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_order_product_lookup: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_order_stats: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_order_tax_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_orders: Data: 0.00MB + Index: 0.02MB + Engine MyISAM
wp_wc_orders_meta: Data: 0.01MB + Index: 0.02MB + Engine MyISAM
wp_wc_product_attributes_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_product_download_directories: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_product_meta_lookup: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_rate_limits: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_reserved_stock: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_tax_rate_classes: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_webhooks: Data: 0.00MB + Index: 0.00MB + Engine MyISAM

### Post Type Counts ###

attachment: 1
page: 7
post: 2
product: 2
shop_order_placehold: 5
wp_navigation: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 8.9.1

### Inactive Plugins (21) ###

Better Search Replace: by WP Engine – 1.4.6
Booking Activities: by Booking Activities Team – 1.16.8
Booking Activities - Advanced Forms: by Booking Activities Team – 1.4.2
Booking Activities - Archiving Tools: by Booking Activities Team – 1.0.0
Booking Activities - Display Pack: by Booking Activities Team – 1.5.3
Booking Activities - Licenses & Updates: by Booking Activities Team – 1.2.10.4
Booking Activities - Notification Pack: by Booking Activities Team – 1.3.5
Booking Activities - Order For Customers: by Booking Activities Team – 1.2.33
Booking Activities - Prices and Credits: by Booking Activities Team – 1.8.10
Booking Activities - Resource Availability: by Booking Activities Team – 1.1.4
Classic Editor: by WordPress Contributors – 1.6.3
OTGS Installer: by OnTheGoSystems – 3.0.4
Payment Plugins for Stripe WooCommerce: by Payment Plugins
support@paymentplugins.com – 3.3.61

qTranslate-XT: by qTranslate Community – 3.15.1
User Role Editor: by Vladimir Garagulya – 4.64.2
W3 Total Cache: by BoldGrid – 2.7.1
WooCommerce Multilingual & Multicurrency: by OnTheGoSystems – 5.3.6
WP Mail SMTP: by WP Mail SMTP – 4.0.1
WPML Media: by OnTheGoSystems – 2.7.3
WPML Multilingual CMS: by OnTheGoSystems – 4.6.7
WPML String Translation: by OnTheGoSystems – 3.2.8

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 89 B

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/
Checkout: #9 - /checkout/
My account: #10 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.1
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
product-pre-publish-modal
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings
async-product-editor-category-field
launch-your-store

Daily Cron: ✔ Next scheduled: 2024-05-30 09:28:32 +00:00
Options: ✔
Notes: 63
Onboarding: skipped

### Action Scheduler ###

Complete: 14
Oldest: 2024-05-27 09:28:49 +0000
Newest: 2024-05-29 14:26:34 +0000

Failed: 1
Oldest: 2024-05-29 14:21:55 +0000
Newest: 2024-05-29 14:21:55 +0000

Pending: 2
Oldest: 2024-05-30 02:19:50 +0000
Newest: 2024-05-30 14:19:50 +0000


### Status report information ###

Generated at: 2024-05-29 14:33:11 +00:00
```

</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes proposed in this Pull Request:

Closes #47967 

This adds an extra validation of the cart server-side to check if it's empty before trying to place an order. This solves an issue we have where on the front-end you could manipulate your cart and remove all items in another window or tab and the Checkout block currently cannot handle updating to reflect that state.

### How to test the changes in this Pull Request:

0. This assumes you're using a block theme (ie Twenty Twenty-Four).

You can try reproduce the bug as listed in the original issue: #47967 
1. Go to a product page
2. Add it to cart
3. Open Cart page in a new tab
4. Open checkout page in a new tab
5. Fill in all the checkout fields but do not place the order
6. Go to the Cart tab and remove the cart item
7. Go to the Checkout page and place the order

Before this used to place an empty order, but now we display an error on the empty checkout page:

Screenshot 2024-10-01 at 9 01 58 PM

I think the look and behaviourUX could be improved in future iterations, but, at least this fixes the bug.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Fix #47967 where user's could place empty orders with the Checkout block.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/StoreApi/Utilities/CartController.php [hunk 1]
```diff
@@ -486,13 +486,29 @@ public function validate_cart() {
...
+		}
+	}
+
 	/**
 	 * Validate all items in the cart and check for errors.
 	 *
 	 * @throws InvalidCartException Exception if invalid data is detected due to insufficient stock levels.
 	 */
 	public function validate_cart_items() {
-		$cart       = $this->get_cart_instance();
```
Comment (samueljseay) [inline plugins/woocommerce/src/StoreApi/Utilities/CartController.php:495]: Unused var, and I don't think we need to call get_cart_instance for any other reason here.

File: plugins/woocommerce/src/StoreApi/Utilities/CartController.php [hunk 2]
```diff
@@ -501,7 +517,7 @@ public function validate_cart_items() {
 		$partial_out_of_stock_products = [];
 		$not_purchasable_products      = [];
 
-		foreach ( $cart_items as $cart_item_key => $cart_item ) {
+		foreach ( $cart_items as $cart_item ) {
```
Comment (samueljseay) [inline plugins/woocommerce/src/StoreApi/Utilities/CartController.php:521]: the key is not used.

File: plugins/woocommerce/src/StoreApi/Utilities/CartController.php [hunk 3]
```diff
@@ -486,13 +486,29 @@ public function validate_cart() {
 		}
 	}
 
+	/**
+	 * When placing an order, validate that the cart is not empty.
+	 *
+	 * @throws InvalidCartException Exception if the cart is empty.
+	 */
+	public function validate_cart_not_empty() {
```
Comment (samueljseay) [inline plugins/woocommerce/src/StoreApi/Utilities/CartController.php:494]: I originally implemented the validation inside `validate_cart_items` but that function also gets called for the cart (which makes sense). I've opted to separate out this validation as it's only relevant when you go to place an order.

File: plugins/woocommerce/src/StoreApi/Utilities/CartController.php (hunk 1 above)
Comment (mikejolley) [reply plugins/woocommerce/src/StoreApi/Utilities/CartController.php:495]: It was likely here to ensure the cart was loaded from session before continuing. It looks like the cart is loaded earlier now, so it should be fine to remove this.

**Review by mikejolley — APPROVED** [review]
Thanks, solves the issue 👍🏻 Only potential problem was the cart instance retrieval (which forces cart session to load) but it appears this is no longer needed and tests are passing.

**Pull request merged**